Receive: implement payment request

rating89us opened the pull request:
This can be used by merchants to create payment requests for products or services, or to create a QR code for asking donations in a website/donation campaign. A QR code and a payment request text (URI) are created containing the data inserted in the fields (both can be copied and easily sent through email, instant messenger, chat, etc.). 

"Description" field is used by Monero GUI wallet as transaction description when the QR code is scanned in Send page. 

"Your name" field is used by Monero GUI wallet when the QR code is scanned in Address book page (user can create a new contact without having to type address/name).

![paymentrequest2](https://user-images.githubusercontent.com/45968869/127735306-09c54f36-b96d-4280-8147-34aa1565a466.gif)

selsta commented:
Here is the promised code for copying an image to clipboard.

```diff
diff --git a/pages/Receive.qml b/pages/Receive.qml
index 666f29fc..96e1982d 100644
--- a/pages/Receive.qml
+++ b/pages/Receive.qml
@@ -153,10 +153,7 @@ Rectangle {
 
                     MenuItem {
                         text: qsTr("Copy to clipboard") + translationManager.emptyString;
-                        onTriggered: {
-                            clipboard.setText(qrCodeText.text);
-                            appWindow.showStatusMessage(qsTr("QR code image copied to clipboard") + translationManager.emptyString, 3);
-                        }
+                        onTriggered: walletManager.saveQrCodeToClipboard(TxUtils.makeQRCodeString(appWindow.current_address))
                     }
 
                     MenuItem {
diff --git a/src/libwalletqt/WalletManager.cpp b/src/libwalletqt/WalletManager.cpp
index 6ecd6b6e..8b04d3b3 100644
--- a/src/libwalletqt/WalletManager.cpp
+++ b/src/libwalletqt/WalletManager.cpp
@@ -31,6 +31,8 @@
 #include "wallet/api/wallet2_api.h"
 #include "zxcvbn-c/zxcvbn.h"
 #include "QRCodeImageProvider.h"
+#include <QClipboard>
+#include <QGuiApplication>
 #include <QFile>
 #include <QFileInfo>
 #include <QDir>
@@ -480,6 +482,14 @@ bool WalletManager::saveQrCode(const QString &code, const QString &path) const
     return QRCodeImageProvider::genQrImage(code, &size).scaled(size.expandedTo(QSize(240, 240)), Qt::KeepAspectRatio).save(path, "PNG", 100);
 }
 
+void WalletManager::saveQrCodeToClipboard(const QString &code) const
+{
+    QClipboard *clipboard = QGuiApplication::clipboard();
+    QSize size;
+    clipboard->setImage(QRCodeImageProvider::genQrImage(code, &size), QClipboard::Clipboard);
+    clipboard->setImage(QRCodeImageProvider::genQrImage(code, &size), QClipboard::Selection);
+}
+
 void WalletManager::checkUpdatesAsync(
     const QString &software,
     const QString &subdir,
diff --git a/src/libwalletqt/WalletManager.h b/src/libwalletqt/WalletManager.h
index 3092a5f3..2b424b73 100644
--- a/src/libwalletqt/WalletManager.h
+++ b/src/libwalletqt/WalletManager.h
@@ -180,6 +180,7 @@ public:
     Q_INVOKABLE bool parse_uri(const QString &uri, QString &address, QString &payment_id, uint64_t &amount, QString &tx_description, QString &recipient_name, QVector<QString> &unknown_parameters, QString &error) const;
     Q_INVOKABLE QVariantMap parse_uri_to_object(const QString &uri) const;
     Q_INVOKABLE bool saveQrCode(const QString &, const QString &) const;
+    Q_INVOKABLE void saveQrCodeToClipboard(const QString &) const;
     Q_INVOKABLE void checkUpdatesAsync(
         const QString &software,
         const QString &subdir,
```

selsta commented on js/TxUtils.js:
```diff
@@ -74,13 +74,29 @@ function isValidOpenAliasAddress(address) {
     }
+    if (txDescription !== undefined && txDescription !== ""){
+        if (amount == ""){
+            qrCodeString += ("?" + XMR_TX_DESCRIPTION + "=" + txDescription)
```
The "spec" says we have to replace spaces with `%20`: https://github.com/monero-project/monero/wiki/URI-Formatting

selsta replied:
You can use `encodeURI(txDescription)` and `encodeURI(recipientName)` here.

selsta reviewed (commented):
```diff
diff --git a/pages/Transfer.qml b/pages/Transfer.qml
index 8646b52e..2e4bc9b6 100644
--- a/pages/Transfer.qml
+++ b/pages/Transfer.qml
@@ -404,7 +404,7 @@ Rectangle {
                                 onTextChanged: {
                                     const parsed = walletManager.parse_uri_to_object(text);
                                     if (!parsed.error) {
-                                        fillPaymentDetails(parsed.address, parsed.payment_id, parsed.amount, parsed.tx_description);
+                                        fillPaymentDetails(parsed.address, parsed.payment_id, parsed.amount, parsed.tx_description, parsed.recipient_name);
                                     }
                                     address = text;
                                 }
```

I would change this too so that recipient name is supported.

selsta commented:
At a later point we can replace the Javascript part with `wallet2::make_uri()` but that still has to be added to wallet2_api.

rating89us commented:
- Fixed spacing in QR code
- Changed description to import recipient + description as "recipient (description)" instead of "recipient description)
- Fixed a bug displaying tooltip when clicking on navbar Payment request button (using Windows):
![receivebug3](https://user-images.githubusercontent.com/45968869/130095240-2bcfa807-5179-44a7-8c1a-932e8d235d5a.gif)
